### ISSUE TITLE
Add JMX beans for IOApp thread pool metrics

### ISSUE THREAD

**Issue by Daenyth** [typelevel/cats-effect#1568]
Spinning this off from #1238 

I think metric collection is an important feature (it is for me, at the very least), so we should enable users to either enrich on their own beans, or build in ones that could be useful.

For example, this is what I'm using right now, building on a standard thread pool:

```scala
  trait ThreadPoolSamplerMBean {
    def getMinThreads: Int
    def getMaxThreads: Int
    def getActiveThreads: Int
    def getCurrentSize: Int
    def getQueuedTasks: Int
  }

  class ThreadPoolSampler(pool: ThreadPoolExecutor) extends ThreadPoolSamplerMBean {
    override def getActiveThreads: Int = pool.getActiveCount
    override def getMinThreads: Int = pool.getCorePoolSize
    override def getMaxThreads: Int = pool.getMaximumPoolSize
    override def getCurrentSize: Int = pool.getPoolSize
    override def getQueuedTasks: Int = pool.getQueue.size
  }
```

And then I register beans of that for both my main compute EC (backing ContextShift) as well as my Blocker EC.

For CE3, it would be *amazing* if we could go even further - build in knowledge of fibers in some way, like number of pending fibers, or number of concurrent fibers waiting for a thread. Worker pool depth, perhaps. I think there's a lot of options there.

**Comment by djspiewak** [comment]
Wdyt, @vasilmkd ?

**Comment by vasilmkd** [comment]
It should be doable. Is there an expectations that these metrics should be dynamic? I'm asking because the thread pool on the JVM uses `Runtime#availableProcessors()`, but they are configured to sleep if there is no work to go around. Again, it should be doable, there's just going to be some discussion as to what metrics are useful and how they should be reported. Otherwise, the the pool is fiber aware, so it can already give you extremely accurate metrics for fibers.

**Comment by Daenyth** [comment]
> Is there an expectations that these metrics should be dynamic

Things like `getActiveThreads`, it's definitely important that it be at least *relatively* up to date. Whether that's updated at the time the method is polled, or independently on a schedule, or constantly updated in a field, doesn't matter so much to me at least

**Comment by vasilmkd** [comment]
Cool, I'm working on a large rewrite of the whole pool which should reduce the code footprint a lot, I've just been busy with other things these weeks. I'd like to get that relatively stable and then I can look into adding configuration for this. It would be an awesome addition, even for us, as we continue to tune and upgrade the pool.

**Comment by djspiewak** [comment]
@vasilmkd Any updated thoughts on this in light of #1609?

**Comment by vasilmkd** [comment]
I have completely missed this comment, I'm sorry. Adding new features to `WorkStealingThreadPool` is done, I'm now working on some code refactoring which will make surfacing this information very easy and straightforward. It's my highest priority to have this in place for stable 3.0.0. No worries.

**Comment by vasilmkd** [comment]
One quick question, where should these MBean interface and class live? Should we have a special subpackage or something?

**Comment by Daenyth** [comment]
I don't have any opinions on that point

**Comment by vasilmkd** [comment]
I'm curious how you expect the MBeans to be instantiated when the only thing we expose through `IORuntime` are `ExecutionContext` instances (and furthermore the blocking EC is not even publicly available). On top of that, the `WorkStealingThreadPool` and the accompanying classes are all internal API. Would we need to expose them for this?

**Comment by djspiewak** [comment]
MBeans in general tend to be instantiated via remote monitoring, not by any internal application access. So I think it's fine to protect the binary interfaces (so they *can't* be instantiated within in-process userspace) and just expose them externally for monitoring purposes. It's all a very dynamic framework, and users would need to be responsible for ensuring that they don't make assumptions in their monitoring and alerting based on `WorkStealingThreadPool` if they're not going to use that pool internally.

**Comment by vasilmkd** [comment]
Closed with #2556.
